storage.go

```diff
@@ -2,35 +2,174 @@ package minio
+		var obj *Object
+		obj, err = objIter.Next()
+		if obj != nil {
+			e := s.client.RemoveObject(ctx, s.bucket, obj.ID, minio.RemoveObjectOptions{})
+			if e != nil {
+				return e
+			}
+		}
+	}
+	if err == IterateDone {
+		return nil
+	}
+	return err
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	input := &objectPageStatus{
+		bufferSize: defaultListObjectBufferSize,
+	}
+
+	options := minio.ListObjectsOptions{WithMetadata: true}
+	switch {
+	case opt.ListMode.IsDir():
+		options.Recursive = false
+		if !strings.HasSuffix(rp, "/") {
+			rp += "/"
+		}
+	case opt.ListMode.IsPrefix():
+		options.Recursive = true
+	default:
+		return nil, services.ListModeInvalidError{Actual: opt.ListMode}
+	}
+	options.Prefix = rp
+
+	input.objChan = s.client.ListObjects(ctx, s.bucket, options)
+
+	return NewObjectIterator(ctx, s.nextObjectPage, input), nil
+}
+
+func (s *Storage) nextObjectPage(ctx context.Context, page *ObjectPage) error {
+	input := page.Status.(*objectPageStatus)
+	for i := 0; i < input.bufferSize; i++ {
+		v, ok := <-input.objChan
+		if !ok {
+			return IterateDone
+		}
+
+		if v.Err == nil {
+			o, err := s.formatFileObject(v)
+			if err != nil {
+				return err
+			}
+			page.Data = append(page.Data, o)
+			input.counter++
+		}
+	}
+	return nil
 }
 
 func (s *Storage) metadata(opt pairStorageMetadata) (meta *StorageMeta) {
-	panic("not implemented")
+	meta = NewStorageMeta()
+	meta.Name = s.bucket
+	meta.WorkDir = s.workDir
+	return meta
 }
 
 func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairStorageRead) (n int64, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	output, err := s.client.GetObject(ctx, s.bucket, rp, minio.GetObjectOptions{})
+	if err != nil {
+		return 0, err
+	}
+	defer output.Close()
+
+	var rc io.ReadCloser = output
+	if opt.HasIoCallback {
+		rc = iowrap.CallbackReadCloser(output, opt.IoCallback)
+	}
+
+	return io.Copy(w, rc)
 }
 
 func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o *Object, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		rp += "/"
+		//TODO
+	}
+
+	output, err := s.client.StatObject(ctx, s.bucket, rp, minio.StatObjectOptions{})
+	if err != nil {
+		return nil, err
+	}
+	o, err = s.formatFileObject(output)
+	if err != nil {
+		return nil, err
+	}
+
+	return
 }
 
 func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int64, opt pairStorageWrite) (n int64, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	r = io.LimitReader(r, size)
+	if opt.HasIoCallback {
+		r = iowrap.CallbackReader(r, opt.IoCallback)
+	}
+
+	options := minio.PutObjectOptions{}
+	if opt.HasContentType {
+		options.ContentType = opt.ContentType
+	}
+	if opt.HasContentMd5 {
+		options.SendContentMd5 = true
```

Comment:
`SendContentMd5` will read content and calculate md5 in memory, we need to read `content-md5` provided by user.

> Specify if you'd like to send content-md5 header with PutObject operation. Note that setting this flag will cause higher memory usage because of in-memory md5sum calculation.

Let's communicate with minio community to have a solution here.

Reply:
ping @aeinrw , I think the problem is not resolved, so I opened it again.

If we can't set `content-md5` header directly via SDK, we need to remove the option here.`

Reply:
Got it！

Reply:
Is there any other question？

Reply:
Mostly LGTM after this fix.